File: pbmx_ffi/java/io/rmf/pbmx/payloads/RandomRevealPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import io.rmf.pbmx.Share;

public final class RandomRevealPayload extends Payload {
    public RandomRevealPayload(String name, Share share, Share.Proof proof) {
        this.name = name;
        this.share = share;
        this.proof = proof;
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.randomReveal(name, share, proof);
    }

    private final String name;
    private final Share share;
    private final Share.Proof proof;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/RandomRevealPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import com.sun.jna.ptr.PointerByReference
import io.rmf.pbmx.Share
import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawShare

class RandomRevealPayload(handle: Pointer) : Payload(handle) {
    val name: String
    val share: Share
    val proof: Share.Proof

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_random_reveal(this.handle, null, length, null, null)
        assert(r == 0)

        val outName = ByteArray(length.value.toInt())
        val outShare = RawShare()
        val outProof = PointerByReference()
        r = FFI.pbmx_unwrap_random_reveal(this.handle, outName, length, outShare, outProof)
        assert(r != 0)
        this.name = outName.toString(Charsets.UTF_8)
        this.share = Share(outShare)
        this.proof = Share.Proof(outProof.value)
    }
}
